### ISSUE TITLE
Nullable field optimization does the Wrong Thing when faced with associated types

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#27532]
## STR

```Rust
use std::mem;
trait Mirror { type Image; }
impl<T> Mirror for T { type Image = T; }
struct S1<T>(<T as Mirror>::Image);
struct S2<T>(T);

fn main() {
    println!("with: {}", mem::size_of::<Option<S1<&u8>>>());
    println!("without: {}", mem::size_of::<Option<S2<&u8>>>());
}
```

## Expected Result

Nullable field optimization should be triggered in both cases

## Result

```
with: 8
without: 16
```

**Comment by jonas-schievink** [comment]
I think you swapped the results by accident

**Comment by eefriedman** [comment]
It looks like the issue is just a couple missing calls to `monomorphize::normalize_associated_type`; I'm working on a fix.

**Comment by arielb1** [comment]
I will fix this with my ADTDef patch (coming real soon now) - I should have made that clearer.
